warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java

```diff
@@ -214,11 +214,11 @@ public synchronized boolean close(ScannerBase bs) {
     }
     
     public synchronized Collection<ScannerBase> currentScanners() {
-        return Collections.unmodifiableSet(instances);
+        return new HashSet<>(instances);
```

Comment:
I understand synchronizing the close call below.  I do not understand why we removed the Collections.unmodifiableSet.

Reply:
Why does it need to be unmodifiable?  We are creating a new set for the caller which ScannerFactory has no ownership of.  Why does the ScannerFactory need to control what the user does with that set?  What the caller does with that set has no impact on ScannerFactory, or the sets contained within.

Reply:
I changed this based on your feedback, but I am changing it back because simply returning an unmodifiable set without creating a new collection _creates_ the potential for a concurrent modification exception as I described in the original issue description, which is exactly what we're trying to avoid.  

As I explained in the issue description, we run into a concurrent modification exception _precisely because_ we are using the exact same set that is used internal to the ScannerFactory.  It _does not matter_ that the set returned is unmodifiable by the caller because it is _still_ modifiable by the ScannerFactory.  So, in ShardQueryLogic, if the ScannerFactory removes an element from either of those sets while we are ShardQueryLogic is iterating through those sets, a concurrent modification exception will be thrown.  This is why we need to return a copy of the scanners and sessions - so that the ScannerFactory can't modify the set while we are iterating through it.

As it pertains to performance, this implementation should actually produce less unnecessary references than the original implementation.  In the original implementation, we wrapped each set with an unmodifiable wrapper, and then in each of the places where this set was used, we created a new list from that collection.  In the current implementation, we have stopped creating the wrapper, and moved the creation of the new list to the call itself, so technically we are dropping one unnecessary reference - the wrapper.

If we are really concerned about the creation of a new list, then my recommendation would be to create synchronized shutdownCurrentScanners and shutdownCurrentSessions methods in the ScannerFactory that can be used to close all of the running scanners and instances in a synchronized fashion.  Note that if we go this route, then we can add back in the unmodifiable wrapper for the currentScanners and currentSessions methods, or just remove those methods _entirely_ since they are only ever used to close the running scanners and instances.

Reply:
I see it now.  Sorry I was not understanding that previously.  Creating a new collection makes sense to me now.